x/ccv/provider/keeper/msg_server.go

```diff
@@ -0,0 +1,73 @@
+package keeper
+
+import (
+	"context"
+
+	cryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	"github.com/cosmos/interchain-security/x/ccv/provider/types"
+	ccvtypes "github.com/cosmos/interchain-security/x/ccv/types"
+	tmstrings "github.com/tendermint/tendermint/libs/strings"
+)
+
+type msgServer struct {
+	*Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the bank MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(keeper *Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+var _ types.MsgServer = msgServer{}
+
+// CreateValidator defines a method for creating a new validator
+func (k msgServer) AssignConsumerKey(goCtx context.Context, msg *types.MsgAssignConsumerKey) (*types.MsgAssignConsumerKeyResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// It is possible to assign keys for consumer chains that are not yet approved.
+	// TODO: In future, a mechanism will be added to limit assigning keys to chains
+	// which are approved or pending approval, only.
+	// Note that current attack potential is restricted because validators must sign
+	// the transaction, and the chainID size is limited.
+
+	providerAddr, err := sdk.ConsAddressFromBech32(msg.ProviderAddr)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
Why is this different from this version?

https://github.com/cosmos/interchain-security/blob/danwt/key-assignment-v4/x/ccv/provider/keeper/msg_server.go#L27-L39

Reply:
Because the code is using `sdk.ConsAddress`, while `ValAddressFromBech32` returns a `sdk.ValAddress`, i.e.,
```golang
// ConsAddress defines a wrapper around bytes meant to present a consensus node.
// When marshaled to a string or JSON, it uses Bech32.
type ConsAddress []byte

// ValAddress defines a wrapper around bytes meant to present a validator's
// operator. When marshaled to a string or JSON, it uses Bech32.
type ValAddress []byte
```

Reply:
`msg.ProviderAddr` is  an operator address of type `ValAddress`
See
https://github.com/cosmos/interchain-security/blob/605d56264a1fbb4cec9c7b2c4a332d7dbcc8418f/x/ccv/provider/client/cli/tx.go#L83

So it shouldn't be converted to a `ConsAddress`

Reply:
Ah, didn't know that. I'll fix it.

Reply:
see https://github.com/cosmos/interchain-security/pull/515/commits/0c9f620017284fea541ac844167c8aaa2a69600a